tests/keyboard/index.ts

```diff
@@ -163,3 +163,30 @@ test('disabling activeElement moves action to HTMLBodyElement', async () => {
     body - keyup: c
   `)
 })
+
+test('typing on focused element with shadow root', async () => {
+  const keypressHandler = mocks.fn()
+  class CustomElement extends HTMLElement {
+    constructor() {
+      super()
+
+      this.attachShadow({mode: 'open'})
+      this.addEventListener('keypress', keypressHandler)
+    }
+  }
+
+  customElements.define('custom-element', CustomElement)
+
+  const {element, user} = setup(
+    '<custom-element tabindex="0"></custom-element>',
+  )
+
+  // Tab to body
+  await user.tab()
+  // Tab to custom element
+  await user.tab()
+  expect(element).toHaveFocus()
+
+  await user.keyboard('[Space]')
+  expect(keypressHandler).toHaveBeenCalled()
```

Comment:
The element (or the first focusable element child) is automatically focused in `setup` (unless `focus: false` option is set). We can simplify the test case.

```suggestion
  const {user, eventWasFired} = setup(
    '<custom-element tabindex="0"></custom-element>',
  )

  await user.keyboard('[Space]')
  expect(eventWasFired('keypress')).toBe(true)
```